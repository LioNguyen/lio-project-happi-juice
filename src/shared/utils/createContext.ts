import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
