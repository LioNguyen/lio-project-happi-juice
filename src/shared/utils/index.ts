export * from './lazyLoad'
export * from './tailwind'
