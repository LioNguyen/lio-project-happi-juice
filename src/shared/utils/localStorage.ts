/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
