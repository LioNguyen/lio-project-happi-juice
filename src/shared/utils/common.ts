import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
