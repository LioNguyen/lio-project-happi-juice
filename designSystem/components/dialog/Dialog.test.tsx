import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import { Dialog, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle, DialogTrigger } from ".";

// Mock RadixUI Dialog components
vi.mock("@radix-ui/react-dialog", () => ({
  Root: vi.fn(({ children }) => <div data-testid="dialog-root">{children}</div>),
  Trigger: vi.fn(({ children }) => <button data-testid="dialog-trigger">{children}</button>),
  Portal: vi.fn(({ children }) => <div data-testid="dialog-portal">{children}</div>),
  Overlay: vi.fn(({ className }) => <div data-testid="dialog-overlay" className={className} />),
  Content: vi.fn(({ children, className }) => (
    <div data-testid="dialog-content" className={className}>
      {children}
    </div>
  )),
  Title: vi.fn(({ children, className }) => (
    <h2 data-testid="dialog-title" className={className}>
      {children}
    </h2>
  )),
  Description: vi.fn(({ children, className }) => (
    <p data-testid="dialog-description" className={className}>
      {children}
    </p>
  )),
  Close: vi.fn(({ children }) => <button data-testid="dialog-close">{children}</button>),
}));

// Mock Lucide icon
vi.mock("lucide-react", () => ({
  X: vi.fn(() => <div data-testid="x-icon" />),
}));

describe("Dialog Component", () => {
  const renderDialog = () => {
    return renderWithTheme(
      <Dialog>
        <DialogTrigger>Open Dialog</DialogTrigger>
        <DialogContent>
          <DialogHeader>
            <DialogTitle>Dialog Title</DialogTitle>
            <DialogDescription>Dialog Description</DialogDescription>
          </DialogHeader>
          <div>Dialog Content</div>
          <DialogFooter>
            <button>Cancel</button>
            <button>Save</button>
          </DialogFooter>
        </DialogContent>
      </Dialog>,
    );
  };

  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderDialog();
    expect(container).toMatchSnapshot();
  });

  // Test basic rendering
  it("should render all dialog components", () => {
    renderDialog();

    expect(screen.getByTestId("dialog-root")).toBeInTheDocument();
    expect(screen.getByTestId("dialog-trigger")).toBeInTheDocument();
    expect(screen.getByTestId("dialog-content")).toBeInTheDocument();
    expect(screen.getByText("Dialog Title")).toBeInTheDocument();
    expect(screen.getByText("Dialog Description")).toBeInTheDocument();
  });

  // Test DialogContent classes
  it("should have correct classes for DialogContent", () => {
    renderDialog();

    const content = screen.getByTestId("dialog-content");
    expect(content).toHaveClass(
      "fixed",
      "left-[50%]",
      "top-[50%]",
      "z-50",
      "grid",
      "w-full",
      "max-w-lg",
      "translate-x-[-50%]",
      "translate-y-[-50%]",
    );
  });

  // Test DialogOverlay classes
  it("should have correct classes for DialogOverlay", () => {
    renderDialog();

    const overlay = screen.getByTestId("dialog-overlay");
    expect(overlay).toHaveClass("fixed", "inset-0", "z-50");
  });

  // Test DialogHeader
  it("should render DialogHeader with correct classes", () => {
    renderDialog();

    const header = screen.getByText("Dialog Title").closest("div");
    expect(header).toHaveClass("flex", "flex-col", "space-y-1.5");
  });

  // Test DialogFooter
  it("should render DialogFooter with correct classes", () => {
    renderDialog();

    const footer = screen.getByText("Cancel").closest("div");
    expect(footer).toHaveClass("flex", "flex-col-reverse", "sm:flex-row", "sm:justify-end", "sm:space-x-2");
  });

  // Test close button
  it("should render close button with X icon", () => {
    renderDialog();

    const closeButton = screen.getByTestId("dialog-close");
    expect(closeButton).toBeInTheDocument();
    expect(screen.getByTestId("x-icon")).toBeInTheDocument();
  });

  // Test title and description styling
  it("should apply correct styles to title and description", () => {
    renderDialog();

    const title = screen.getByTestId("dialog-title");
    const description = screen.getByTestId("dialog-description");

    expect(title).toHaveClass("text-lg", "font-semibold", "leading-none", "tracking-tight");
    expect(description).toHaveClass("text-sm", "text-muted-foreground");
  });

  // Test accessibility
  it("should have appropriate ARIA attributes", () => {
    renderDialog();

    const closeButton = screen.getByTestId("dialog-close");
    expect(closeButton.querySelector(".sr-only")).toHaveTextContent("Close");
  });

  // Test dialog trigger interaction
  it("should call RadixUI Dialog.Root when trigger is clicked", async () => {
    const user = userEvent.setup();
    renderDialog();

    const trigger = screen.getByTestId("dialog-trigger");
    await user.click(trigger);

    // Since we're using a mock, we can verify that the Root component was rendered
    expect(screen.getByTestId("dialog-root")).toBeInTheDocument();
  });

  // Test animation states
  it("should have correct animation classes", () => {
    renderDialog();

    const content = screen.getByTestId("dialog-content");
    expect(content).toHaveClass(
      "data-[state=open]:animate-in",
      "data-[state=closed]:animate-out",
      "data-[state=closed]:fade-out-0",
      "data-[state=open]:fade-in-0",
    );
  });
});
