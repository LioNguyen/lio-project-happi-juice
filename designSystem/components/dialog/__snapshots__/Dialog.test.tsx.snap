// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog Component > should match snapshot with default props 1`] = `
<div>
  <div
    data-testid="dialog-root"
  >
    <button
      data-testid="dialog-trigger"
    >
      Open Dialog
    </button>
    <div
      data-testid="dialog-portal"
    >
      <div
        class="fixed inset-0 z-50 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
        data-testid="dialog-overlay"
      />
      <div
        class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg"
        data-testid="dialog-content"
      >
        <div
          class="flex flex-col space-y-1.5 text-center sm:text-left"
        >
          <h2
            class="text-lg font-semibold leading-none tracking-tight"
            data-testid="dialog-title"
          >
            Dialog Title
          </h2>
          <p
            class="text-sm text-muted-foreground"
            data-testid="dialog-description"
          >
            Dialog Description
          </p>
        </div>
        <div>
          Dialog Content
        </div>
        <div
          class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
        >
          <button>
            Cancel
          </button>
          <button>
            Save
          </button>
        </div>
        <button
          data-testid="dialog-close"
        >
          <div
            data-testid="x-icon"
          />
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
